### PR TITLE
Add test to verify that lowering gpus on restart works with sharded spawn

### LINKED ISSUE #14485: Can sharded be resumed with a different number of devices?
## 🚀 Feature

Does this work?

```
model = BoringModel()
trainer = Trainer(strategy="ddp_sharded_spawn", fast_dev_run=True, gpus=2)
trainer.fit(model)

checkpoint_path = os.path.join(tmpdir, "model.pt")
trainer.save_checkpoint(checkpoint_path)

model = BoringModel()
trainer = Trainer(strategy="ddp_sharded_spawn", fast_dev_run=True, gpus=1)
trainer.fit(model, ckpt_path=checkpoint_path)
```

### Motivation

We had a legacy test skipped in our CI mentioning that this is unsupported.

### Pitch

We should check, and raise an appropriate error if that's the case.

### Additional context

https://github.com/Lightning-AI/lightning/pull/14476/files/38e10ba837dad423ecaa52f300609083de379e19#r960749044
______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging PyTorch Lightning, Transformers, and Hydra.

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #14485

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @awaelchli @rohitgr7 @akihironitta

### CODE REVIEW

File: tests/tests_pytorch/strategies/test_sharded_strategy.py [hunk 1]
```diff
@@ -18,6 +20,37 @@
...
+            self._is_equal(actual, expected) for actual, expected in zip(self.state_dict(), self.old_model_state_dict)
+        )
+
+        for optimizer, state in zip(self.trainer.optimizers, self.old_optimizer_states):
+            optimizer.consolidate_state_dict()
+            self._is_equal(optimizer.state_dict(), state)
+
+    def _is_equal(self, a, b):
+        if isinstance(a, torch.Tensor):
+            return torch.all(torch.eq(a, b))
```
Comment (justusschock) [inline tests/tests_pytorch/strategies/test_sharded_strategy.py:46]:
```suggestion
            return torch.allclose(a, b)
```

otherwise this might fail due to minor numeric inaccuracies. `allclose` == `eq` for integer tensors and `allclose` == `a - b < eps` for floating point :)

File: tests/tests_pytorch/strategies/test_sharded_strategy.py [hunk 2]
```diff
@@ -18,6 +20,37 @@
...
+
+        if isinstance(a, Mapping):
```
Comment (justusschock) [inline tests/tests_pytorch/strategies/test_sharded_strategy.py:48]: would it be easier to use `apply_to_collections` here with `torch.allclose` on every `Tensor`?

**Review by justusschock — APPROVED** [review]

File: tests/tests_pytorch/strategies/test_sharded_strategy.py [hunk 3]
```diff
@@ -18,6 +20,37 @@
...
+        return optimizer
+
+
+class CheckModelRestore(ModelWithAdamOptimizer):
+    def __init__(self, old_model_state_dict, old_optimizer_states):
+        super().__init__()
+        self.old_model_state_dict = old_model_state_dict
+        self.old_optimizer_states = old_optimizer_states
+
+    def on_train_start(self):
```
Comment (awaelchli) [inline tests/tests_pytorch/strategies/test_sharded_strategy.py:35]: We discussed this briefly in the past offline in general. 

We will have to provide a standardized way to do this in Lightning. Not necessarily automatically on resuming, but at least by providing utilities in some form to do this, not just for Zero1, but for all types of sharded state dicts and checkpoints. 

I will look into making a GitHub issue.

File: tests/tests_pytorch/strategies/test_sharded_strategy.py [hunk 4]
```diff
@@ -18,6 +20,37 @@
...
+        assert all(
+            self._is_equal(actual, expected) for actual, expected in zip(self.state_dict(), self.old_model_state_dict)
+        )
+
+        for optimizer, state in zip(self.trainer.optimizers, self.old_optimizer_states):
+            optimizer.consolidate_state_dict()
+            self._is_equal(optimizer.state_dict(), state)
```
Comment (awaelchli) [inline tests/tests_pytorch/strategies/test_sharded_strategy.py:42]: I do not understand why this is necessary. 

When we checkpoint the states, we already consolidate the state dict. When the state gets loaded for the single device, it already is consolidated. And optimizer.consolidate_state_dict should be a no-op anyway on single device. 

Please remind me where I am wrong here.

Comment (rohitgr7) [reply tests/tests_pytorch/strategies/test_sharded_strategy.py:42]: looks like one can call `state_dict(all_ranks=True)` to do that.
https://github.com/facebookresearch/fairscale/blob/ee647b976cf4c8fdd37bc9ae3fd6331d225ba2a0/fairscale/optim/oss.py#L400-L404

since it relies on `self._all_states` with is an empty list by default.

File: tests/tests_pytorch/strategies/test_sharded_strategy.py (hunk 3 above)
Comment (awaelchli) [reply tests/tests_pytorch/strategies/test_sharded_strategy.py:35]: Apart from deepspeed, we don't support saving a shard of a checkpoint on each rank in any of the sharded strategyes. Issues like this one for example exist #11361. Once we support that, we will need to provide ways of resharding these checkpoints in an efficient way. 

Imagine you have a shard of N checkpoint files because you trained with N processes. Now you want to load these checkpoints into a trainer that runs with M != N processes.

The inefficient way we do it currently is to call consolidate_state_dict, but this will have to hold all shard in memory which is often not possible with the biggest models today.

File: tests/tests_pytorch/strategies/test_sharded_strategy.py (hunk 4 above)
Comment (awaelchli) [reply tests/tests_pytorch/strategies/test_sharded_strategy.py:42]: No, there must be a misunderstanding here. You set devices=1, there is only one rank. [consolidation](https://pytorch.org/docs/stable/distributed.optim.html#torch.distributed.optim.ZeroRedundancyOptimizer.consolidate_state_dict) is essentially a no-op here. Can you explain why you did it?

Comment (rohitgr7) [reply tests/tests_pytorch/strategies/test_sharded_strategy.py:42]: yes, I understand that theoretically, consolidate call is not required here but that's not how fairscale is designed. it relies on `self._all_states` to generate the `state_dict`
https://github.com/facebookresearch/fairscale/blob/ee647b976cf4c8fdd37bc9ae3fd6331d225ba2a0/fairscale/optim/oss.py#L305

File: tests/tests_pytorch/strategies/test_sharded_strategy.py (hunk 3 above)
Comment (rohitgr7) [reply tests/tests_pytorch/strategies/test_sharded_strategy.py:35]: yep! agreed!

but is that blocking this PR? I understand we need to consolidate it first, so once we have a better way, we could update this test.

if you think we should close it, it's fine too :)

File: tests/tests_pytorch/strategies/test_sharded_strategy.py [hunk 5]
```diff
@@ -321,3 +354,33 @@ def test_ddp_sharded_strategy_checkpoint_multi_gpu_fairscale_optimizer(tmpdir, s
     # Assert model parameters are identical after loading
     for trained_param, loaded_param in zip(model.parameters(), saved_model.parameters()):
         assert torch.equal(trained_param.to("cpu"), loaded_param)
+
+
+@RunIf(min_cuda_gpus=2, fairscale=True)
+def test_ddp_sharded_strategy_fit_ckpt_path_downsize_gpus(tmpdir):
```
Comment (awaelchli) [inline tests/tests_pytorch/strategies/test_sharded_strategy.py:360]: Note: This test takes ~17 seconds to finish.

File: tests/tests_pytorch/strategies/test_sharded_strategy.py (hunk 3 above)
Comment (awaelchli) [reply tests/tests_pytorch/strategies/test_sharded_strategy.py:35]: No it's not blocking. It's just making aware that we are looking into options to generalize this and make it memory efficient.

File: tests/tests_pytorch/strategies/test_sharded_strategy.py (hunk 5 above)
Comment (rohitgr7) [reply tests/tests_pytorch/strategies/test_sharded_strategy.py:360]: oh! that's too much. wondering if we should create another one `slow-gpu` so these could run separately.

Comment (rohitgr7) [reply tests/tests_pytorch/strategies/test_sharded_strategy.py:360]: cc: @carmocca @Borda

File: tests/tests_pytorch/strategies/test_sharded_strategy.py (hunk 2 above)
Comment (rohitgr7) [reply tests/tests_pytorch/strategies/test_sharded_strategy.py:48]: well, we'd still need to check other types which are not tensors and are part of this `Mapping`.

File: tests/tests_pytorch/strategies/test_sharded_strategy.py (hunk 5 above)
Comment (carmocca) [reply tests/tests_pytorch/strategies/test_sharded_strategy.py:360]: The current workflow could be ported to run on Azure with GPUs. This wasnt done in the past because the GPU bill was expensive, but now that it runs on our cluster it should be okay. I suggest you open an issue

Comment (rohitgr7) [reply tests/tests_pytorch/strategies/test_sharded_strategy.py:360]: done #15372

**Review by Borda — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
